### PR TITLE
Fix #11692: Improve doc comments within macros

### LINKED ISSUE #11692: Improve doc comments inside macros
## Summary

Some items defined within macros use macro tokens like `$UDF` within their doc comments, which render as that literal string in cargo documentation, which is confusing and looks a little broken. Here's an example:

https://docs.rs/datafusion/latest/datafusion/functions/math/fn.cot.html

## Proposal

Use a combination of `#[doc]`, `concat!` and `stringify!` to properly interpolate relevant type names into the generated doc comments.

### PR DESCRIPTION
**Pull request by Rafferty97**
## Which issue does this PR close?

Closes #11692.

## Rationale for this change

Marginally improves the quality of the public-facing documentation.

## What changes are included in this PR?

- Fix the way doc comments are generated for certain macro-generated items

## Are these changes tested?

I have manually tested that the resulting doc comments now appear correctly by running `cargo doc`.

## Are there any user-facing changes?

Yes, some doc comments now show the correct types rather than generic strings like "$UDF".

### CODE REVIEW

File: datafusion/functions-nested/src/macros.rs [hunk 1]
```diff
@@ -90,9 +90,8 @@ macro_rules! create_func {
             #[allow(non_upper_case_globals)]
             static [< STATIC_ $UDF >]: std::sync::OnceLock<std::sync::Arc<datafusion_expr::ScalarUDF>> =
                 std::sync::OnceLock::new();
-            /// ScalarFunction that returns a [`ScalarUDF`] for [`$UDF`]
-            ///
-            /// [`ScalarUDF`]: datafusion_expr::ScalarUDF
+
+            #[doc = concat!("ScalarFunction that returns a [`ScalarUDF`](datafusion_expr::ScalarUDF) for [`", stringify!($UDF), "`]")]
```
Comment (alamb) [inline datafusion/functions-nested/src/macros.rs:94]: ❤️

**Review by alamb — COMMENTED** [review]
Thank you @Rafferty97 -- I have noticed this too

**Comment by alamb** [comment]
I kicked off CI -- once that passes I think this PR will be good to go

**Comment by Rafferty97** [comment]
It looks like `cargo doc` failed with this error:

```
error: unresolved link to `r`
  --> datafusion/functions/src/macros.rs:78:17
   |
78 |         #[doc = concat!("Return a [`ScalarUDF`](datafusion_expr::ScalarUDF) for [`", stringify!($UDF), "`]")]
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
  ::: datafusion/functions/src/core/mod.rs:41:1
   |
41 | make_udf_function!(r#struct::StructFunc, STRUCT, r#struct);
   | ---------------------------------------------------------- in this macro invocation
   |
   = note: the link appears in this line:
           
           Return a [`ScalarUDF`](datafusion_expr::ScalarUDF) for [`r#struct::StructFunc`]
                                                                   ^^^^^^^^^^^^^^^^^^^^^^
   = note: no item named `r` in scope
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
   = note: this error originates in the macro `make_udf_function` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I've investigated it but I can't figure out why this is occurring (other than the raw identifier clearly being an issue) or how to remediate it. If I type out the same doc string manually, like so:

```
/// Return a [`ScalarUDF`](datafusion_expr::ScalarUDF) for [`r#struct::StructFunc`]
pub fn foo() {}
```

and run `cargo doc` locally, it works exactly as expected.

**Comment by alamb** [comment]
I played around with this PR a bit (as I think docs are quite important) -- thanks @Rafferty97  

I found I could fix the link in the docs like this:
```rust
        #[doc = "Return a [`ScalarUDF`](datafusion_expr::ScalarUDF) for [`"]
        #[doc = stringify!($UDF)]
        #[doc = "`]"]
        pub fn $NAME() -> std::sync::Arc<datafusion_expr::ScalarUDF> {
```

However, then I got warning about documenting private items. So I chose to just skip the links 

```
warning: public documentation for `cardinality_udf` links to private item `Cardinality`
  --> datafusion/functions-nested/src/macros.rs:94:21
   |
94 |               #[doc = concat!("ScalarFunction that returns a [`ScalarUDF`](datafusion_expr::ScalarUDF) for [`", stringify!($UDF), "`]")]
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
  ::: datafusion/functions-nested/src/cardinality.rs:31:1
   |
31 | / make_udf_expr_and_func!(
32 | |     Cardinality,
33 | |     cardinality,
34 | |     array,
35 | |     "returns the total number of elements in the array.",
36 | |     cardinality_udf
37 | | );
   | |_- in this macro invocation

```

**Review by alamb — APPROVED** [review]
Thanks again @Rafferty97  -- please let me know if you are happy with the changes I made to this PR to get the CI check passing

**Comment by Rafferty97** [comment]
@alamb - Yes, I trust your judgement. Happy to merge :)

**Comment by alamb** [comment]
I think we can always make docs better, this is a step in the right direction. Thanks again.

**Pull request merged**